src/pudl/etl/epacems_assets.py

```diff
@@ -148,8 +149,438 @@ def hourly_emissions_epacems(
+) -> pd.DataFrame:
+    """Make unique annual plant_id_eia and emissions_unit_id_epa.
+
+    TODO: is there a better/faster way to do this?
+
+    Returns:
+        dataframe with unique set of: "plant_id_eia", "report_year" and
+        "emissions_unit_id_epa"
+    """
+    epacems_ids = (
+        hourly_emissions_epacems[["plant_id_eia", "year", "emissions_unit_id_epa"]]
+        .drop_duplicates()
+        .compute()
+    )
+
+    return epacems_ids
+
+
+@asset(io_manager_key="pudl_sqlite_io_manager")
+def epacamd_eia_subplant_ids(
+    clean_epacamd_eia: pd.DataFrame,
+    generators_eia860: pd.DataFrame,
+    emissions_unit_ids_epacems: pd.DataFrame,
+    boiler_generator_assn_eia860: pd.DataFrame,
+) -> pd.DataFrame:
+    """Groups units and generators into unique subplant groups.
+
+    This function consists of three primary parts:
+
+    #.  Augement the EPA CADM:EIA crosswalk with all IDs from
+    #.  Use graph analysis to identify distinct groupings of EPA units and EIA
+        generators based on 1:1, 1:m, m:1, or m:m relationships.
+
+
+    Returns:
+        table of cems_ids and with subplant_id added
+    """
+    clean_epacamd_eia = (
+        augement_crosswalk_with_eia860(clean_epacamd_eia, generators_eia860)
+        .pipe(augement_crosswalk_with_epacamd_ids, emissions_unit_ids_epacems)
+        .pipe(augement_crosswalk_with_bga_eia860, boiler_generator_assn_eia860)
+    )
+    # filter the crosswalk to drop any units that don't exist in CEMS
+    # use graph analysis to identify subplants
+    subplant_crosswalk_complete = pudl.analysis.epacamd_eia.make_subplant_ids(
+        clean_epacamd_eia
+    ).pipe(pudl.metadata.fields.apply_pudl_dtypes, "eia")
+    # update the subplant_crosswalk to ensure completeness
+    # prepare the subplant crosswalk by adding a complete list of generators and adding the unit_id_pudl column
+
+    # TODO: Should we migrate the unit_id_pudl complete fleshing out from outputs
+    # -> transform? The unit_id_pudl is NOT annually varying.
+
+    # update the subplant ids for each plant
+    subplant_crosswalk_complete = subplant_crosswalk_complete.groupby(
+        by=["plant_id_eia"], group_keys=False
+    ).apply(update_subplant_ids)
+    # remove the intermediate columns created by update_subplant_ids
+    subplant_crosswalk_complete["subplant_id"].update(
+        subplant_crosswalk_complete["new_subplant"]
+    )
+    subplant_crosswalk_complete = subplant_crosswalk_complete.reset_index(
+        drop=True
+    ).drop_duplicates(
+        subset=[
+            "plant_id_epa",
+            "emissions_unit_id_epa",
+            "plant_id_eia",
+            "generator_id",
+            "subplant_id",
+        ],
+        keep="last",
+    )
+    return subplant_crosswalk_complete
+
+
+def augement_crosswalk_with_eia860(
+    crosswalk_clean: pd.DataFrame, generators_eia860: pd.DataFrame
+) -> pd.DataFrame:
+    """Merge any plants that are missing from the EPA crosswalk but appear in EIA-860.
```

Comment:
Good eye! But from the docs in `clean_epacamd_eia` is looks like we only do this for the sake of the fast CI tests! It looks like the filtering only happens then because of FK constraints.